onnxruntime/core/providers/cpu/tensor/where_op.cc

```diff
@@ -0,0 +1,118 @@
+#include "core/providers/cpu/tensor/where_op.h"
+
+#include "core/providers/cpu/math/element_wise_ops.h"  // for broadcast utilities
+
+namespace onnxruntime {
+
+// kernel builder functions
+#define WHERE_TYPED_KERNEL_WITH_TYPE_NAME(type, type_name)                         \
+  ONNX_CPU_OPERATOR_TYPED_KERNEL(                                                  \
+      Where,                                                                       \
+      9,                                                                           \
+      type_name,                                                                   \
+      KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<type>()), \
+      Where<type>)
+
+#define WHERE_TYPED_KERNEL(type) \
+  WHERE_TYPED_KERNEL_WITH_TYPE_NAME(type, type)
+
+// start with a subset of types, enable more as needed...
+//WHERE_TYPED_KERNEL(uint8_t)
+//WHERE_TYPED_KERNEL(uint16_t)
+//WHERE_TYPED_KERNEL(uint32_t)
+//WHERE_TYPED_KERNEL(uint64_t)
+//WHERE_TYPED_KERNEL(int8_t)
+//WHERE_TYPED_KERNEL(int16_t)
+//WHERE_TYPED_KERNEL(int32_t)
+//WHERE_TYPED_KERNEL(int64_t)
+//WHERE_TYPED_KERNEL(MLFloat16)
+//WHERE_TYPED_KERNEL(BFloat16)
+WHERE_TYPED_KERNEL(float)
+//WHERE_TYPED_KERNEL(double)
+WHERE_TYPED_KERNEL_WITH_TYPE_NAME(std::string, string)
```

Comment:
Not sure how well Eigen supports a scalar type of std::string. It seems to work from basic testing, but perhaps it would be safer to use a span instead for the string version of this op. Anyone know more about this?

Reply:
@KeDengMS may have insights here :).

Reply:
Added non-Eigen implementation.